x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractScalarFunctionTestCase.java

```diff
@@ -604,9 +604,9 @@ private static Stream<List<DataType>> allPermutations(int argumentCount) {
         if (argumentCount > 3) {
             throw new IllegalArgumentException("would generate too many combinations");
         }
-        Stream<List<DataType>> stream = representable().map(t -> List.of(t));
+        Stream<List<DataType>> stream = DataType.validFunctionParameters().map(t -> List.of(t));
```

Comment:
We're filtering types just to reduce the amount of test cases? Or is there another reason?

Reply:
Lots of reasons!

* When we implement a new type we sometimes have functions that declare they work on "any type" but they don't work on the new types. We should implement those functions on those types. Like equality, TO_STRING, MV_FIRST, and stuff.
* Some types are never supported and there's other infrastructure to check that - like `UNSUPPORTED` and `DOC_DATA_TYPE` and `PARTIAL_AGG`. The numeric widening types too.

I wanted a place where I could explain *why* we're filtering each type with a little comment.

Reply:
I feel like this is a blacklist, and we should be using a whitelist instead. Declaring in each function which types they can really consume (We do it already, to some extent).
The fact that this is a blacklist just for testing tells me that we could be actually using one of those types in a function, and because of this, we wouldn't know, as the test won't check it (?)

Reply:
I think we should remove almost everything from this list. I like it being a blocklist so new types get opted-in by default. If you make a new type and you want to exclude it from testing you can do that here and mark it with a comment like `// new type, implementations coming` or something.

I want to try removing `SOURCE` and `TSID` from this list because those are real types that we really pass into the functions. Nothing supports them. But that's fine. We'll test the errors. I just want to do it in a follow-up.